src/Util/Input.cpp

```diff
@@ -92,6 +119,8 @@ void Input::Update() {
     s_MouseState[Keycode::MOUSE_MB].first =
         s_MouseState[Keycode::MOUSE_MB].second;
 
+    s_Io = ImGui::GetIO();
+
     while (SDL_PollEvent(&s_Event) != 0) {
```

Comment:
I think you should block the inputs here instead of the getters.

Reply:
how would I modify the logic in `IsKeyXXX`?

Reply:
I'm not sure if there are any unnoticed side effects, but this is how I would approach it.
```patch
diff --git a/src/Util/Input.cpp b/src/Util/Input.cpp
index a741fb4..4748ab8 100644
--- a/src/Util/Input.cpp
+++ b/src/Util/Input.cpp
@@ -28,12 +28,9 @@ ImGuiIO Input::s_Io; // allocate memory only because it is invalid
                      // to call `ImGui::GetIO()` at this time
 
 bool Input::IsKeyPressed(const Keycode &key) {
-    if (s_Io.WantCaptureMouse || s_Io.WantTextInput) {
-        return false;
-    }
 
     if (key > Keycode::NUM_SCANCODES) {
-        return s_MouseState[key].second; // && !s_MouseState[key].first;
+        return s_MouseState[key].second;
     }
 
     const auto index = static_cast<const int>(key);
@@ -41,9 +38,6 @@ bool Input::IsKeyPressed(const Keycode &key) {
 }
 
 bool Input::IsKeyDown(const Keycode &key) {
-    if (s_Io.WantCaptureMouse || s_Io.WantTextInput) {
-        return false;
-    }
 
     if (key > Keycode::NUM_SCANCODES) {
         return s_MouseState[key].second && !s_MouseState[key].first;
@@ -54,9 +48,6 @@ bool Input::IsKeyDown(const Keycode &key) {
 }
 
 bool Input::IsKeyUp(const Keycode &key) {
-    if (s_Io.WantCaptureMouse || s_Io.WantTextInput) {
-        return false;
-    }
 
     if (key > Keycode::NUM_SCANCODES) {
         return !s_MouseState[key].second && s_MouseState[key].first;
@@ -67,17 +58,11 @@ bool Input::IsKeyUp(const Keycode &key) {
 }
 
 bool Input::IsMouseMoving() {
-    if (s_Io.WantCaptureMouse) {
-        return false;
-    }
 
     return s_MouseMoving;
 }
 
 bool Input::IfScroll() {
-    if (s_Io.WantCaptureMouse) {
-        return false;
-    }
 
     return s_Scroll;
 }
@@ -87,15 +72,13 @@ bool Input::IfExit() {
 }
 
 glm::vec2 Input::GetScrollDistance() {
-    if (s_Io.WantCaptureMouse) {
-        return {0.0F, 0.0F};
-    }
 
     return s_ScrollDistance;
 }
 
 void Input::Update() {
-    int x, y;
+    int x;
+    int y;
     SDL_GetMouseState(&x, &y);
     s_CursorPosition.x = static_cast<float>(x);
     s_CursorPosition.y = static_cast<float>(y);
@@ -122,6 +105,10 @@ void Input::Update() {
     s_Io = ImGui::GetIO();
 
     while (SDL_PollEvent(&s_Event) != 0) {
+        if (s_Io.WantCaptureMouse || s_Io.WantTextInput) {
+            ImGui_ImplSDL2_ProcessEvent(&s_Event);
+            continue;
+        }
         if (s_Event.type == SDL_MOUSEBUTTONUP &&
             s_Event.button.button == SDL_BUTTON_LEFT) {
             s_MouseState[Keycode::MOUSE_LB].second = false;
@@ -157,10 +144,6 @@ void Input::Update() {
         }
         s_MouseMoving = s_Event.type == SDL_MOUSEMOTION || s_MouseMoving;
         s_Exit = s_Event.type == SDL_QUIT;
-
-        // ? I think this is bad, but IDK if we should move `s_Event` to
-        // ? `Core::Context` and use callback functions
-        ImGui_ImplSDL2_ProcessEvent(&s_Event);
     }
 }
 
```